### PR TITLE
Add at-root-no-redundant rule

### PR DESCRIPTION
**Pull request by pamelalozano16**
Fix (#601).

### CODE REVIEW

File: src/rules/at-root-no-redundant-selector/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant-selector/index.js:14]: You don't have to manually go down the tree recursively, you can use `walkAtRules` with any container node, which inclides AtRule nodes, so.. you can do this instead:

```js
function findAtRoot(node) {
  let atRoot;
  node.walkAtRules('at-root', node => {
    atRoot = node;
    return false;
  });
  return atRoot;
}
```

File: src/rules/at-root-no-redundant-selector/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant-selector/index.js:61]: You don't want to report the error on the `@keyframes`, instead you want to report it on the `@at-rule` node.

You can implement the suggestion from above with `findAtRoot`, and report it there, e.g.:
   ```js
   if (node.name === "keyframes") {
     const atRoot = findAtRoot(node);
     if (atRoot) {
       utils.report({
         message: messages.rejected,
         node: atRoot,
         result,
         ruleName
       });
     }
   }
   ```

File: src/rules/at-root-no-redundant-selector/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant-selector/index.js:44]: Optional: you can pass a filter to `walkAtRules`, for example:

```js
root.walkAtRules('at-root', node => { ... })
```

Doing so, you can omit the checks for `node.name === 'at-root'` because it would be guaranteed to be this node. However, you won't be able to match `@keyframes`. But.. you can always iterate up, for example, if I want to check whether an at-root exists within a `@keyframes` node, I could do:

```js
function isWithinKeyframes(node) {
  let parent = node.parent;
  while (parent != null) {
    if (parent.type === 'at-rule' && parent.name === 'keyframes') {
      return true;
    }
    parent = node.parent;
  }
  return false;
}
```

With this, your overall implementation could be greatly reduced to:

```js
function rule(actual) {
  return (root, result) => {
    const validOptions = utils.validateOptions(result, ruleName, {
      actual
    });

    if (!validOptions) {
      return;
    }

    root.walkAtRules('at-root', node => {
      if (
        node.parent.type === 'root' ||
        node.params.replace(/#{.*}/g, '').includes('&') ||
        isWithinKeyframes(node)) {
          utils.report({
            message: messages.rejected,
            node,
            result,
            ruleName
          });
        }
    });
}
```

**Review by Goodwine — CHANGES_REQUESTED** [review]

File: src/rules/at-root-no-redundant-selector/README.md
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/README.md:1]: I had initially suggested `no-redundant-at-root` as well, @nex3 suggested `at-root-no-redundant-selector`, thoughts? 🤔

File: src/rules/at-root-no-redundant-selector/__tests__/index.js
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/__tests__/index.js:13]: Oh right, I'll verify all of them thanks!! :)

File: src/rules/at-root-no-redundant-selector/__tests__/index.js
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/__tests__/index.js:19]: Oh right, I was searching for an interpolation with any character instead of an interpolation with &.

File: src/rules/at-root-no-redundant-selector/index.js
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/index.js:14]: Great suggestion, thank you!!

File: src/rules/at-root-no-redundant-selector/README.md
Comment (Goodwine) [reply src/rules/at-root-no-redundant-selector/README.md:1]: Maybe we can discuss this today at coworking time, I still think that `no-redundant-at-root` is a more accurate name, since the selector is not the problem, even less so when the `@at-root` is already at the root

Comment (Goodwine) [reply src/rules/at-root-no-redundant-selector/README.md:1]: discussed this briefly, the id should indeed start with `at-root` following the same pattern as other checks, so I think then the ID should be `at-root-no-redundant`, without the "selector" part 

ultimately the final say will be from the maintainer :)

File: src/rules/at-root-no-redundant-selector/index.js
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/index.js:49]: Solved this by matching the regex to `/#{[&]}/g`

File: src/rules/at-root-no-redundant-selector/index.js
Comment (pamelalozano16) [reply src/rules/at-root-no-redundant-selector/index.js:44]: Awesome!! Thanks

File: src/rules/at-root-no-redundant/README.md
Comment (Goodwine) [inline src/rules/at-root-no-redundant/README.md:24]:
```suggestion
.a { @at-root .b & { margin: 3px; } }
```

File: src/rules/at-root-no-redundant/README.md
Comment (Goodwine) [inline src/rules/at-root-no-redundant/README.md:49]:
```suggestion
.a { @at-root .b#{&} { margin: 3px; } }
```

File: src/rules/at-root-no-redundant/__tests__/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant/__tests__/index.js:56]: Nit: this doesn't have to have any contents to assert the behavior being tested, it can be simplified to `@mixin foo {}`

File: src/rules/at-root-no-redundant/__tests__/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant/__tests__/index.js:134]:
```suggestion
      }`,
```

File: src/rules/at-root-no-redundant/index.js
Comment (Goodwine) [inline src/rules/at-root-no-redundant/index.js:11]: Nit: this condition is a little bit risky as it could throw an NPE if the node is "orphan". In practice it probably doesn't matter, but a safer check would be `while(parent)` or to be more explicit `while(parent != null)`

**Review by Goodwine — APPROVED** [review]
LGTM besides the remaining comments

**Comment by pamelalozano16** [comment]
@kristerkari please let us know your thoughts :)

**Comment by kristerkari** [comment]
Thanks! I'll need to go through all the new PRs 👍

File: src/rules/at-root-no-redundant/README.md
Comment (kristerkari) [inline src/rules/at-root-no-redundant/README.md:3]: There are some small things to point out related to how the README and message should be formatted. 

Please check Stylelint's developer guide for more info:
https://stylelint.io/developer-guide/rules#write-the-readme

```md
The single-line description is in the form of:

    - "Disallow ..." for no rules
    - "Limit ..." for max rules
    - "Require ..." for rules that accept "always" and "never" options
    - "Specify ..." for everything else
```

File: src/rules/at-root-no-redundant/index.js
Comment (kristerkari) [inline src/rules/at-root-no-redundant/index.js:21]: If I'm not mistaken this should be `Unexpected @at-root rule`.

https://stylelint.io/developer-guide/rules#add-problem-messages

Comment (pamelalozano16) [reply src/rules/at-root-no-redundant/index.js]: Right, thank you!

**Review by kristerkari — APPROVED** [review]

**Pull request merged**
